pkg/storageops/gce/gce.go

```diff
@@ -0,0 +1,540 @@
+package gce
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"regexp"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+	"golang.org/x/oauth2/google"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/libopenstorage/openstorage/pkg/storageops"
+	compute "google.golang.org/api/compute/v1"
+)
+
+var notFoundRegex = regexp.MustCompile(`.*notFound`)
+
+const googleDiskPrefix = "/dev/disk/by-id/google-"
+
+type gceOps struct {
+	inst    *instance
+	service *compute.Service
+	mutex   sync.Mutex
+}
+
+// instance stores the metadata of the running GCE instance
+type instance struct {
+	ID         string
+	Name       string
+	Hostname   string
+	Zone       string
+	Project    string
+	InternalIP string
+	ExternalIP string
+	LBRequest  string
+	ClientIP   string
+	Error      string
+}
+
+type assigner struct {
+	err error
+}
+
+func (a *assigner) assign(getVal func() (string, error)) string {
+	if a.err != nil {
+		return ""
+	}
+
+	s, err := getVal()
+	if err != nil {
+		a.err = err
+	}
+
+	return s
+}
+
+func getEnvValueStrict(key string) (val string, err error) {
+	if val = os.Getenv(key); len(val) != 0 {
+		return
+	}
+
+	err = fmt.Errorf("env variable %s is not set", key)
+	return
+}
+
+// gceInfo fetches the GCE instance metadata from the metadata server
+func gceInfo(inst *instance) {
+	a := &assigner{}
```

Comment:
This probably should not be a single instance. `assigner` overwrites the error returned if two or more errors were saved. This causes the overwritten errors to be lost.

Is there a need for `assign`? Could we remove it? Maybe replace the whole thing with a for loop and struct of the variable to save to and the function to get the information from.

Reply:
assigner doesn't override errors as the assign method checks for existing errors before invoking the next function.

```
func (a *assigner) assign(getVal func() (string, error)) string {
	if a.err != nil {
		return ""
	}
```

Reply:
Correct, but the value of `a.err` is lost and instead an empty string is returned. It should at least log the error before returning and if it logs, it should log the name it was trying to assign.